Preparing for Fastboot Sandbox

taras opened the pull request:
The goal of this PR is to make it easier to use content generated by `ember-writer` in Fastboot Sandbox mode by making json output JSONAPI spec compliant. This will allow us to fetch and store data from the sandbox in JSONAPI format without doing any formatting in the browser.

taras committed:
item-counts no longer counts empty strings

taras committed:
Data is now being serialized to JSONAPI

taras committed:
API namespace is now configurable

taras committed:
Authors are now frontmatter md files

- Removed /data/authors.js from genertor
- Added /authors directory to generator
- Adding .md file to /authors will parse it same as articles

taras committed:
Renamed all articles to posts

cball commented on blueprints/author/index.js:
```diff
@@ -0,0 +1,32 @@
+const moment = require('moment');
+
+module.exports = {
+  description: 'Generates a new blog article',
```
author?

taras replied:
done

cball commented on index.js:
```diff
@@ -50,8 +49,8 @@ module.exports = EngineAddon.extend({
     let blogFiles = new Funnel(this.blogDirectory, {
-      destDir: 'api/blog',
-      include: ['*.md']
+      destDir: this.addonConfig.namespace,
```
👍

cball reviewed (commented):
@taras this is a great start, thanks! I'll spend some time on Friday reworking our current use-case to use this format and make sure all is well with both approaches before we merge this in.

I like the serializer idea, will think through that a bit more too.

cball commented:
Also note we are getting the following test failure. Seems like we need to update the author fixture data.

```
Chris Ball is an author of a post but is not a known author. Expecting to find an author with slug Chris Ball
Error: Chris Ball is an author of a post but is not a known author. Expecting to find an author with slug Chris Ball
    at _array.uniq.map.name (/home/ubuntu/ember-writer/index.js:109:15)
    at Array.map (native)
    at Class._visibleAuthors (/home/ubuntu/ember-writer/index.js:105:33)
    at Class.postBuild (/home/ubuntu/ember-writer/index.js:68:21)
    at /home/ubuntu/ember-writer/node_modules/ember-cli/lib/models/builder.js:206:34
    at Array.map (native)
    at Class.processAddonBuildSteps (/home/ubuntu/ember-writer/node_modules/ember-cli/lib/models/builder.js:204:43)
    at tryCatch (/home/ubuntu/ember-writer/node_modules/rsvp/dist/lib/rsvp/-internal.js:215:12)
    at invokeCallback (/home/ubuntu/ember-writer/node_modules/rsvp/dist/lib/rsvp/-internal.js:230:13)
    at publish (/home/ubuntu/ember-writer/node_modules/rsvp/dist/lib/rsvp/-internal.js:198:7)
    at flush (/home/ubuntu/ember-writer/node_modules/rsvp/dist/lib/rsvp/asap.js:85:5)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

taras committed:
Corrected author blueprint description

taras commented:
To fix the failing tests, I need to update the API requests to match JSONAPI response format.

Instead, I could add the missing Ember Data bits to make it work properly in with Ember Data. I can do this for the demo app first and then we can see how to integrate into the addon. What do you think?

taras committed:
Big rewrite

1. Using Fixturify to parse data directory and store all attributes in memory
2. Use in-memory data to build relationships for posts, tags and authors
3. All relationships are now working
4. All relationships are sideloaded for all endpoints

taras commented:
@cball I rewrote a bunch of the addon plugin and all of the blog parser code. 

The problem with previous implementation is that it used a filter to convert all of the `.md` files to `.json`. This worked fine for simple use cases but for sideloading data, we need all of the data available in the serializer.

The new code uses [`node-fixturify`](https://github.com/joliss/node-fixturify) to read the data directory. I'm reading all of the data in the beginning, processing it in ES6 getters and outputting a tree using [broccoli-fixturify](https://github.com/rwjblue/broccoli-fixturify).

I updated the [dummy app](https://github.com/this-dot/ember-writer/tree/fastboot-sandbox-ftw/tests/dummy) to show the new setup using Ember Data.

It's a pretty big change, so I want to make sure that you agree with this direction before updating the node tests.

taras committed:
Fixed: blog not rebuilding when file changes

cball commented:
Hey @taras I took a look at this and it looks like a good direction. I'm giving it a go on our internal app today and will report back

taras commented:
FYI, you need a tags directory and authors. This makes it obvious what tags
and authors are available. I can make that optional if necessary

-- 
Taras Mankovski

On December 5, 2016 at 8:43:59 PM, Chris Ball (notifications@github.com)
wrote:

> Hey @taras <https://github.com/taras> I took a look at this and it looks
> like a good direction. I'm giving it a go on our internal app today and
> will report back
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/echobind/ember-writer/pull/33#issuecomment-264879581>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEjv1BsBlbzXU4p3sLCvj7QWdRF2xDsks5rFCo3gaJpZM4K6C3i>
> .
>

taras committed:
Adapters should be inferred from application adapter

taras committed:
Added Page type

taras committed:
Upgraded to Ember 2.12 and removed engines
